### PR TITLE
[SPARK-16441][YARN] Set maxNumExecutor depends on yarn cluster resources.

### PR DESCRIPTION
**Pull request by wangyum**
## What changes were proposed in this pull request?

Dynamic set `spark.dynamicAllocation.maxExecutors` by cluster resources.

## How was this patch tested?

manual test and unit test

### CODE REVIEW

**Comment by srowen** [comment]
I don't think this is a necessary change. Already, you can't ask for more resources than the cluster has; the cluster won't grant them. Capping it here means the app can't use more resources if the cluster suddenly gets more.

I see the problem you're trying to solve but the resource manager already ramps up requests slowly, so I don't think this is the issue.

**Comment by andrewor14** [comment]
I agree. Resource managers generally expect applications to request more than what's available already so we don't have to do it again ourselves in Spark.

**Comment by wangyum** [comment]
It will reduce the function call on [CoarseGrainedSchedulerBackend.requestTotalExecutors()](https://github.com/apache/spark/blob/v2.1.0/core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala#L491) after apply this PR:
[before](https://github.com/apache/spark/files/756685/doNotApply-PR-16819.txt) [after-apply-this-PR](https://github.com/apache/spark/files/756684/apply-PR-16819.txt)

Full log can be found [here](https://issues.apache.org/jira/secure/attachment/12851318/SPARK-16441-compare-apply-PR-16819.zip).

**Comment by srowen** [comment]
What problem does this solve though? calling that function is not a problem. It seems like you get the right behavior in both cases. Are you saying there's some RPC problem? The target goes very high, but, as far as I can see it's correctly reflecting the fact that the app would use a lot of executors if it could -- that's fine.

**Comment by wangyum** [comment]
@srowen . Dynamic set `spark.dynamicAllocation.maxExecutors`  can avoid some strange problems:

1. [Spark application hang when dynamic allocation is enabled](https://issues.apache.org/jira/browse/SPARK-16441)
2. [Report failure reason from Reporter Thread](https://issues.apache.org/jira/browse/SPARK-19226)
3. CLI shows successful but web ui didn't, simally to [this](https://issues.apache.org/jira/secure/attachment/12846513/can-not-consume-taskEnd-events.jpg)

I add a unit test just now.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 1]
```diff
@@ -1193,6 +1189,37 @@ private[spark] class Client(
...
+   * then set the max number of executors depends on yarn cluster VCores Total.
+   * If not using dynamic allocation don't set it.
+   */
+  private def setMaxNumExecutors(): Unit = {
+    if (Utils.isDynamicAllocationEnabled(sparkConf)) {
+
+      val defaultMaxNumExecutors = DYN_ALLOCATION_MAX_EXECUTORS.defaultValue.get
+      if (defaultMaxNumExecutors == sparkConf.get(DYN_ALLOCATION_MAX_EXECUTORS)) {
+        val executorCores = sparkConf.getInt("spark.executor.cores", 1)
+        val maxNumExecutors = yarnClient.getNodeReports().asScala.
```
Comment (Devian-ua) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: Shouldn't we take queue's maxResources amount into account from [ResourceManager REST APIs](https://hadoop.apache.org/docs/r2.7.2/hadoop-yarn/hadoop-yarn-site/ResourceManagerRest.html)?

Comment (wangyum) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: Good suggestion. I will try API first. Pseudo code:
```
import org.apache.hadoop.yarn.client.api.{YarnClient, YarnClientApplication}
import scala.collection.JavaConverters._
import org.apache.hadoop.yarn.api.protocolrecords._
import org.apache.hadoop.yarn.api.records._
import org.apache.hadoop.yarn.conf.YarnConfiguration

val yarnConf = new YarnConfiguration()
val yarnClient = YarnClient.createYarnClient
yarnClient.init(yarnConf)
yarnClient.start()
yarnClient.getRootQueueInfos
```

**Comment by vanzin** [comment]
I agree there's room for improvement in the current code; I even asked SPARK-18769 to be filed to track that work.

But I don't think setting the max to a fixed value at startup is the right approach. Queue configs change, node managers go up and down, new ones are added, old ones are removed. If this value ends up being calculated at the wrong time, the application will suffer. If you want to investigate a more dynamic approach here I'm all for that, but I'm not a big fan of the current solution.

**Comment by wangyum** [comment]
@vanzin We must pull the configuration from ResourceManager, ResourceManager can't push.
So setting the max before each stage? This feels too frequent.

In fact, This is suitable for periodic tasks. e.g. ML, SQL,
For streaming jobs, it is better set manually.

**Comment by vanzin** [comment]
Getting the config only at the beginning, to me, is not an acceptable solution.

Getting it every once in a while is better, but it's not the only possible approach. I even suggest something different in the bug I mention above.

**Comment by tgravescs** [comment]
I agree with others, this is not the way to do this. There are different schedulers in yarn, each with different configs that could affect the actual resources you get. 

If you want to do something like this it should look at the available resources after calling the allocate call to yarn (allocateResponse.getAvailableResources).  When yarn returns it tells you the available resources, which takes into account  the various scheduler things. 

MapReduce refers to that as headroom and uses it to determine things like if it needs to kill a reducer to run a map.  We could use this to help with dynamic allocation and do more intelligent things.

**Comment by wangyum** [comment]
@vanzin What do you think about current approach? I have tested on a same Spark hive-thriftserver, the `spark.dynamicAllocation.maxExecutors` wiil decrease if I kill 4 NodeManager:
```
17/02/27 15:58:08 DEBUG ExecutorAllocationManager: Not adding executors because our current target total is already 94 (limit 94)
17/02/27 15:58:09 DEBUG ExecutorAllocationManager: Not adding executors because our current target total is already 94 (limit 94)
17/02/27 16:05:49 DEBUG ExecutorAllocationManager: Not adding executors because our current target total is already 85 (limit 85)
17/02/27 16:05:49 DEBUG ExecutorAllocationManager: Not adding executors because our current target total is already 85 (limit 85)
```

**Comment by vanzin** [comment]
So your current approach is to have a second connection to the RM, and ask for the RM's available resources every time the scheduler tries to change the number of resources.

Did you look at Tom's suggestion of using {{AllocateResponse.getAvailableResources()}} instead? Seems like it would be simpler, cheaper, and could all be handled internally in {{YarnAllocator.scala}}.

**Pull request closed without merging**